crates/bevy_render/src/render_resource/pipeline_cache.rs

```diff
@@ -325,220 +210,115 @@ impl LayoutCache {
     }
 }
 
-/// Cache for render and compute pipelines.
-///
-/// The cache stores existing render and compute pipelines allocated on the GPU, as well as
-/// pending creation. Pipelines inserted into the cache are identified by a unique ID, which
-/// can be used to retrieve the actual GPU object once it's ready. The creation of the GPU
-/// pipeline object is deferred to the [`RenderStage::Render`] stage, just before the render
-/// graph starts being processed, as this requires access to the GPU.
-///
-/// Note that the cache do not perform automatic deduplication of identical pipelines. It is
-/// up to the user not to insert the same pipeline twice to avoid wasting GPU resources.
-///
-/// [`RenderStage::Render`]: crate::RenderStage::Render
-#[derive(Resource)]
-pub struct PipelineCache {
-    layout_cache: LayoutCache,
-    shader_cache: ShaderCache,
-    device: RenderDevice,
-    pipelines: Vec<CachedPipeline>,
-    waiting_pipelines: HashSet<CachedPipelineId>,
-    new_pipelines: Mutex<Vec<CachedPipeline>>,
+#[derive(Copy, Clone, Hash, Eq, PartialEq, Debug)]
+enum CachedPipelineId {
+    Render(RenderPipelineId),
+    Compute(ComputePipelineId),
 }
 
-impl PipelineCache {
-    pub fn pipelines(&self) -> impl Iterator<Item = &CachedPipeline> {
-        self.pipelines.iter()
+impl From<RenderPipelineId> for CachedPipelineId {
+    fn from(value: RenderPipelineId) -> Self {
+        Self::Render(value)
+    }
+}
+
+impl From<ComputePipelineId> for CachedPipelineId {
+    fn from(value: ComputePipelineId) -> Self {
+        Self::Compute(value)
     }
+}
 
-    /// Create a new pipeline cache associated with the given render device.
-    pub fn new(device: RenderDevice) -> Self {
+struct CachedPipeline<I, D, P> {
+    pub id: I,
```

Comment:
These don't need to be pub if the struct isn't pub.

Reply:
done